src/semaphore.c

```diff
@@ -716,6 +785,10 @@ _dispatch_thread_semaphore_create(void)
 	int ret = sem_init(s4, 0, 0);
 	DISPATCH_SEMAPHORE_VERIFY_RET(ret);
 	return (_dispatch_thread_semaphore_t)s4;
+#elif USE_FUTEX_SEM
```

Comment:
not reall related to that very code but:

`_dispatch_get_thread_semaphore` and `_dispatch_put_thread_semaphore` are meant to amortize the cost of the syscall to create a kernel semaphore.

on linux where a futex is used, caching the malloc() is not useful, as we allocate a continuation which has its own cache.

`_dispatch_put_thread_semaphore` and `_dispatch_get_thread_semaphore` should be the same as `dispose` and `create` when USE_FUTEX_SEM is in use like this:

```C
#if USE_FUTEX_SEM
#define _dispatch_get_thread_semaphore() _dispatch_thread_semaphore_create()
#define _dispatch_put_thread_semaphore() _dispatch_thread_semaphore_dispose()
#else
DISPATCH_ALWAYS_INLINE
static inline _dispatch_thread_semaphore_t
_dispatch_get_thread_semaphore(void)
{
	_dispatch_thread_semaphore_t sema = (_dispatch_thread_semaphore_t)
			_dispatch_thread_getspecific(dispatch_sema4_key);
	if (slowpath(!sema)) {
		return _dispatch_thread_semaphore_create();
	}
	_dispatch_thread_setspecific(dispatch_sema4_key, NULL);
	return sema;
}

DISPATCH_ALWAYS_INLINE
static inline void
_dispatch_put_thread_semaphore(_dispatch_thread_semaphore_t sema)
{
	_dispatch_thread_semaphore_t old_sema = (_dispatch_thread_semaphore_t)
			_dispatch_thread_getspecific(dispatch_sema4_key);
	_dispatch_thread_setspecific(dispatch_sema4_key, (void*)sema);
	if (slowpath(old_sema)) {
		return _dispatch_thread_semaphore_dispose(old_sema);
	}
}
#endif
```